### PR TITLE
Inserter: Prevent non-deterministic order of inserter items

### PR DESCRIPTION
**Pull request by fluiddot**
**This PR takes the changes created by @ttahmouch from PR https://github.com/WordPress/gutenberg/pull/34062, thanks for the help on this ❤️!**

## Description

We've noticed that using the function `Array.prototype.sort` can produce different results on the native version, actually on Android due to the use of [Hermes](https://hermesengine.dev/), it's generating a different list than iOS (which uses [`javascriptcore`](https://developer.apple.com/documentation/javascriptcore)). Per the documentation of the function, it looks like [the function itself is not stable until ES2019](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#sort_stability) so it would be nice to figure out a replacement to prevent non-deterministic results.

Specifically, we've identified this issue in the `getInserterItems` selector ([reference](https://github.com/WordPress/gutenberg/blob/b77066b38ae7e7a8894b33f328ccfeedfd265c8f/packages/block-editor/src/store/selectors.js#L1636-L1639)), so a solution has been applied for this purpose by replacing the use of `sort` with the calculation of the core and non-core block lists using `reduce`. This directly affects the logic related to prioritizing core blocks' display in the inserter menu.

### Unit test

Block variations case has been added to the unit test "getInserterItems with core blocks prioritization" to assure the proper order of blocks in the inserter menu.

#### Verify that inserter items order match the production version

For this purpose, we can compare the current order of a WP.com site with a local instance of Gutenberg using the changes from this PR. 

1. Run Gutenberg locally using the changes from this PR.
2. Open a post/page.
3. Open the inserter menu by clicking on ➕ button located in the top toolbar.
Screenshot 2021-08-16 at 19 35 19

4. In a different tab, navigate to a WP.com site.
5. Open a post/page.
6. Open the inserter menu by clicking on ➕ button located in the top toolbar.
7. Compare both inserter menus (local instance and WP.com site) and verify that are displaying the inserter items in the same order.

**NOTE:** In a WP.com site the inserter menu shows more items than a local instance, however, the extra items should always be displayed at the end of the list.

Additionally, we should verify that the inserter menu displayed within the blocks matches the same order as the blocks are registered.

1. Run Gutenberg locally using the changes from this PR.
2. Open a post/page on incognito mode (**NOTE:** Incognito mode is required to prevent displaying suggested items).
3. Select a block and open the inserter menu by clicking on ➕ button.
Screenshot 2021-08-16 at 19 35 14

4. Verify that the inserter items are sorted in the same order as the blocks are registered.
https://github.com/WordPress/gutenberg/blob/34fbec666d9b210e94df56246bc208da64d3ad10/packages/block-library/src/index.js#L119-L188

### Native version

Similar to the web version, we have to verify that the inserter items order remains the same as in the production version, in this case, we would need to use the iOS version because the Android one is not displaying the correct order. Additionally, we need to verify that both iOS and Android versions display the same order.

#### Verify that inserter items order match the production version

1. Download and install the last version of the WordPress app from the Appstore on an iOS device.
2. Open a post/page.
3. Open the inserter menu by tapping on ➕ button.
4. (On a different iOS device) Open the app and use the changes from this PR by connecting it to a local Metro server.
5. Open a post/page.
6. Open the inserter menu by tapping on ➕ button.
7. Compare both inserter menus (iOS production version and iOS local version) and verify that are displaying the inserter items in the same order.

### Verify IOS and Android match the inserter items order

1. Open the app on iOS.
2. Open a post/page.
3. Open the inserter menu by tapping on ➕ button.
4. (On an Android device) Open the app.
5. Open a post/page.
6. Open the inserter menu by tapping on ➕ button.
7. Compare both inserter menus (iOS and Android) and verify that are displaying the inserter items in the same order.
8. Verify that the inserter items are sorted in the same order as the blocks are registered.
https://github.com/WordPress/gutenberg/blob/947816e3d5867d8285f1151054d576b1d76811f1/packages/block-library/src/index.native.js#L237-L269

**NOTE:** The following blocks are not displayed in the inserter menu:
- `missing`
- `column`
- `classic`
- `button`
- `socialLink`
- `reusableBlock`

The block `core/search` is displayed at the bottom because it's being included as a core variant.
https://github.com/WordPress/gutenberg/blob/947816e3d5867d8285f1151054d576b1d76811f1/packages/block-library/src/search/variations.js#L6-L12

## Screenshots 

iOS|Android
--|--
|

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Comment by ttahmouch** [comment]
The test seems reasonable to me, @fluiddot .

File: packages/block-editor/src/store/selectors.js
Comment (dcalhoun) [inline packages/block-editor/src/store/selectors.js]: Nitpick: renaming this might better communicate the intent. Feel free to ignore this comment, however. 
```suggestion
		const groupByType = ( blocks, block ) => {
```

**Review by dcalhoun — COMMENTED** [review]
Well done! This is a creative yet straightforward solution to this odd problem. Thank you. This looks good to me. 

I verified each of the test cases outlined in the PR description. I left one comment regarding naming, but it is not a blocker from my perspective. 

I would imagine we should wait for a web team member's input before merging, so I will yield approval to them.

File: packages/block-editor/src/store/selectors.js
Comment (fluiddot) [reply packages/block-editor/src/store/selectors.js]: Thanks for the suggestion @dcalhoun 🙇 , I've just applied it in [this commit](https://github.com/WordPress/gutenberg/pull/34078/commits/a72b5abcfc9694d27005a824a757a02feedf016a).

**Comment by twstokes** [comment]
LGTM @fluiddot!

Tested:

- [x]  Web
- [x] iOS
- [x] Android (with and without [Hermes](https://github.com/facebook/hermes) in debug mode)

**Comment by mchowning** [comment]
Thanks for this fix @fluiddot and @ttahmouch ! 🙇  

I wonder if there's room for us to improve the default block order (in a different PR). This jumped out at me because the "more" block didn't seem to make sense in the third position (ahead of, for example, image, video, and list). Any thoughts on this @iamthomasbishop ? I know that we're planning to revamp the inserter soon, but if the order can be improved it would be a pretty easy fix.

Related previous discussion: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1105

File: packages/block-editor/src/store/selectors.js
Comment (ttahmouch) [reply packages/block-editor/src/store/selectors.js]: I named it `toTyped` not because I particularly liked that name, but because I like for array operations to read like prose as much as possible, e.g., `blockVariations.reduce( toTyped )`. I just couldn't think of a good name when I put up the draft PR. The name wasn't the focus at the time, but rather the nature of the solution.

I don't feel particularly strongly about changing it again, but maybe something like `blockVariations.reduce( toGroupedBlockTypes )` would be more expressive?

Also, I believe `groupByType` can be moved to the module lexical scope instead of the local functional scope because I don't believe I closured anything in the function. I intended to do that in the draft PR and forgot.

**Comment by ttahmouch** [comment]
> Thanks for this fix @fluiddot and @ttahmouch ! 🙇
> 
> I wonder if there's room for us to improve the default block order (in a different PR). This jumped out at me because the "more" block didn't seem to make sense in the third position (ahead of, for example, image, video, and list). Any thoughts on this @iamthomasbishop ? I know that we're planning to revamp the inserter soon, but if the order can be improved it would be a pretty easy fix.
> 
> Related previous discussion: [wordpress-mobile/gutenberg-mobile#1105](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1105)

I think I agree with the sentiment behind this. I suppose I assumed the order would be self-correcting based on the frecency. However, it never quite made sense to me when attempting the dis/allow list implementation why the default order of the block types differed between [`web`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/index.js#L149) and [`native`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/index.native.js#L241), and why the entry point module for registering the block types in `native` [exported a default list of the block types](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/index.native.js#L70) but didn't acknowledge them when actually attempting to register them.

File: packages/block-editor/src/store/selectors.js
Comment (dcalhoun) [reply packages/block-editor/src/store/selectors.js]: > [...] I like for array operations to read like prose as much as possible, e.g., `blockVariations.reduce( toTyped )`. 

> I don't feel particularly strongly about changing it again, but maybe something like `blockVariations.reduce( toGroupedBlockTypes )` would be more expressive?

Ah, I see. I did not read or recognize that. That does make sense and a name like `toGroupedBlockTypes` could express that well. I'm fine with whichever name @fluiddot prefers. 👍🏻

**Comment by gziolo** [comment]
> I wonder if there's room for us to improve the default block order (in a different PR). This jumped out at me because the "more" block didn't seem to make sense in the third position (ahead of, for example, image, video, and list). Any thoughts on this @iamthomasbishop ? I know that we're planning to revamp the inserter soon, but if the order can be improved it would be a pretty easy fix.

The order is different for the list of most common blocks between web and mobile, so it seems correct how those blocks are presented in the inserter. However, I agree with the sentiment that they should be presented in the same order on both platforms.

File: packages/block-editor/src/store/test/selectors.js
Comment (gziolo) [inline packages/block-editor/src/store/test/selectors.js:3792]: Good catch with id, here 👍🏻

**Review by gziolo — APPROVED** [review]
All tests pass so we should be good 😄

File: packages/block-editor/src/store/selectors.js
Comment (fluiddot) [reply packages/block-editor/src/store/selectors.js]: I don't have a strong opinion on the naming either but since this PR should be merged as soon as possible as we're aiming to include it in the next release, I'm leaning towards preventing more changes.

> Also, I believe groupByType can be moved to the module lexical scope instead of the local functional scope because I don't believe I closured anything in the function. I intended to do that in the draft PR and forgot.

I think it's ok to keep it there, the [previous function `prioritizeCoreBlocks`](https://github.com/WordPress/gutenberg/blob/20d9c54f6684627486c9e59a4a24848b4bd4f4ef/packages/block-editor/src/store/selectors.js#L1621-L1630) was defined in the same place and its use is mostly exclusive to the block type sorting logic.

**Comment by fluiddot** [comment]
> Thanks for this fix @fluiddot and @ttahmouch ! 🙇
> 
> I wonder if there's room for us to improve the default block order (in a different PR). This jumped out at me because the "more" block didn't seem to make sense in the third position (ahead of, for example, image, video, and list). Any thoughts on this @iamthomasbishop ? I know that we're planning to revamp the inserter soon, but if the order can be improved it would be a pretty easy fix.
> 
> Related previous discussion: [wordpress-mobile/gutenberg-mobile#1105](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1105)

The default block order in the native version is driven by the block registration order defined here:
https://github.com/WordPress/gutenberg/blob/20d9c54f6684627486c9e59a4a24848b4bd4f4ef/packages/block-library/src/index.native.js#L236-L270

And on the web version is here:
https://github.com/WordPress/gutenberg/blob/20d9c54f6684627486c9e59a4a24848b4bd4f4ef/packages/block-library/src/index.js#L119-L189

I'm wondering if we could unify them 🤔 , in the meantime and as a quick fix, we could update the list on the native version to match the same order as in the web.

**Comment by fluiddot** [comment]
This PR has approvals for both versions native ([approval #1](https://github.com/WordPress/gutenberg/pull/34078#pullrequestreview-730900145) and [approval #2](https://github.com/WordPress/gutenberg/pull/34078#issuecomment-899701711)) and web ([approval #3](https://github.com/WordPress/gutenberg/pull/34078#pullrequestreview-731448436)), so it's ready to be merged.

**Comment by gziolo** [comment]
@fluiddot, it looks like the code for registering core blocks could be unified between the native mobile and web now that most of the blocks work on both platforms. A good example would be to extract common blocks (the group in the web version) to a new file and import them in both versions of `index.js`. This way any further changes will be reflected everywhere.

**Comment by fluiddot** [comment]
> @fluiddot, it looks like the code for registering core blocks could be unified between the native mobile and web now that most of the blocks work on both platforms. A good example would be to extract common blocks (the group in the web version) to a new file and import them in both versions of `index.js`. This way any further changes will be reflected everywhere.

That's a great idea, in the native version we would still require to filter out the unsupported blocks but I agree that extracting the common blocks will make the order consistent between web and native 🎊 .

**Comment by gziolo** [comment]
You can also keep some of the imported blocks behind the `Platform.is( 'web' )` check if that helps.

**Comment by fluiddot** [comment]
> You can also keep some of the imported blocks behind the `Platform.is( 'web' )` check if that helps.

Yep, that would help for sure. Let's try this option and see how it looks like.

**Comment by youknowriad** [comment]
Looking at the performance graph for the "inserter hover" metric, it seems it started to be a bit less performant around this commit https://codehealth.vercel.app 

While it's not certain 100%, it sounds like this commit is the most likely in that range of commits to impact it. The increase is not big enough to be very concerning but might be a good indicator to at least try to debug/check.

**Comment by fluiddot** [comment]
> Looking at the performance graph for the "inserter hover" metric, it seems it started to be a bit less performant around this commit https://codehealth.vercel.app
> 
> While it's not certain 100%, it sounds like this commit is the most likely in that range of commits to impact it. The increase is not big enough to be very concerning but might be a good indicator to at least try to debug/check.

That's interesting, after checking the graph (see attached screenshot), I see that the commit that introduced this change `8c12930` is indeed in that range. It would be nice to run some checks with a different approach in case it helps towards the performance.

Screenshot 2021-08-19 at 12 45 24

**Pull request merged**
